### PR TITLE
Manage resource and onscreen contexts using separate IOSGLContext objects

### PR DESCRIPTION
**Pull request by gw280**
Attempt 3. So the issue here was that with the change to WeakPtr for IOSGLContext we ended up with a nullptr deref on simulator where there is no GL context.

I've fixed that and verified it passes LUCI: 

https://chromium-swarm.appspot.com/task?id=4744ea5e1a2d3010#
https://chromium-swarm.appspot.com/task?id=4745214d8f1a6610

I am also going to file a follow up issue to clean up the logic surrounding `[FlutterView createSurfaceWithResourceGLContext]` as it's quite confusing right now with the way it deals with simulator and device.

### CODE REVIEW

File: shell/platform/darwin/ios/platform_view_ios.mm [hunk 1]
```diff
@@ -42,16 +42,24 @@
...
     accessibility_bridge_.reset();
   }
   owner_controller_ = owner_controller;
   if (owner_controller_) {
-    ios_surface_ =
-        [static_cast<FlutterView*>(owner_controller.get().view) createSurface:gl_context_];
+    fml::WeakPtr<IOSGLContext> weak_gl_context;
+    if (resource_gl_context_) {
+      weak_gl_context = resource_gl_context_->GetWeakPtr();
+    }
```
Comment (gw280) [inline shell/platform/darwin/ios/platform_view_ios.mm]: This is what changed.

**Comment by gw280** [comment]
Update to fix the failing unit test that was caught by @xster's new test (thanks!)

So basically what was going on is that this patch released the IOSSurface in order to not retain the IOSurface when backgrounded (a requirement of the issue that this patch was to resolve). This happened in PlatformViewIOS::NotifyDestroyed(). When the app came back into the foreground, though, FlutterViewController::surfaceUpdated would call NotifyCreated which would then try and create a GPU surface but the IOSSurface was gone, thus returning a nullptr and ending up with a nullptr deref down the chain (in the rasterizer).

This patch recreates the IOSSurface correctly when coming into the foreground, providing the owner FlutterViewController is still present.

**Comment by xster** [comment]
Another scenario we should validate for (tests being added in https://github.com/flutter/engine/pull/12128) is to not unconditionally re-claim resources on foregrounding. 

For instance, if I showed a VC, popped it but held reference to it, then background and foregrounded the app, I shouldn't recreate the IOSSurface again until I present that VC again.

**Comment by gw280** [comment]
I think I'd consider that an optimisation that's out of scope for this PR, let's file a new one.

**Pull request merged**